# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [2]:
import pandas as pd

## imports for Python, Pandas

In [3]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [4]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [5]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [6]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population info.governor    state shortname
0        Dade       12345    Rick Scott  Florida        FL
1     Broward       40000    Rick Scott  Florida        FL
2  Palm Beach       60000    Rick Scott  Florida        FL
3      Summit        1234   John Kasich     Ohio        OH
4    Cuyahoga        1337   John Kasich     Ohio        OH

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [7]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{u'_id': {u'$oid': u'52b213b38594d8a2be17c780'},
  u'approvalfy': 1999,
  u'board_approval_month': u'November',
  u'boardapprovaldate': u'2013-11-12T00:00:00Z',
  u'borrower': u'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  u'closingdate': u'2018-07-07T00:00:00Z',
  u'country_namecode': u'Federal Democratic Republic of Ethiopia!$!ET',
  u'countrycode': u'ET',
  u'countryname': u'Federal Democratic Republic of Ethiopia',
  u'countryshortname': u'Ethiopia',
  u'docty': u'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  u'envassesmentcategorycode': u'C',
  u'grantamt': 0,
  u'ibrdcommamt': 0,
  u'id': u'P129828',
  u'idacommamt': 130000000,
  u'impagency': u'MINISTRY OF EDUCATION',
  u'lendinginstr': u'Investment Project Financing',
  u'lendinginstrtype': u'IN',
  u'lendprojectcost': 550000000,
  u'majorsector_percent': [{u'Name': u'Education', u'Percent': 46},
   {u'Name': u'Education', u'Percent': 26},
   {u'Name': u'Public Administration, Law, and Ju

In [8]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

_id  approvalfy board_approval_month  \
0  {u'$oid': u'52b213b38594d8a2be17c780'}        1999             November   
1  {u'$oid': u'52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia   
1          TN                      Republic of Tunisia          Tunisia   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   

                                              theme1  \
0   {u'Percent': 100, u'Name': u'Education for all'}   
1  {u'Percent': 30, u'Name': u'Other economic man...   

                                      theme_namecode themecode   totalamt  \
0  [{u'code': u'65', u'name': u'Education for all'}]        65  130000000   
1  [{u'code': u'24', u'name': u'Other economic ma...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

### 0. Loading and Exploring Data

In [9]:
df = pd.read_json('data/world_bank_projects.json')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

- 500 projects (rows) each with 50 features (columns)
- There are some missing columns
- 7 integer types and 43 object data types

In [11]:
# check the first 10 lines
df.head(10)

_id  approvalfy board_approval_month  \
0  {u'$oid': u'52b213b38594d8a2be17c780'}        1999             November   
1  {u'$oid': u'52b213b38594d8a2be17c781'}        2015             November   
2  {u'$oid': u'52b213b38594d8a2be17c782'}        2014             November   
3  {u'$oid': u'52b213b38594d8a2be17c783'}        2014              October   
4  {u'$oid': u'52b213b38594d8a2be17c784'}        2014              October   
5  {u'$oid': u'52b213b38594d8a2be17c785'}        2014              October   
6  {u'$oid': u'52b213b38594d8a2be17c786'}        2014              October   
7  {u'$oid': u'52b213b38594d8a2be17c787'}        2014              October   
8  {u'$oid': u'52b213b38594d8a2be17c788'}        2014              October   
9  {u'$oid': u'52b213b38594d8a2be17c789'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   
5  2013-10-31T00:00:00Z                        REPUBLIC OF KENYA   
6  2013-10-29T00:00:00Z                      GOVERNMENT OF INDIA   
7  2013-10-29T00:00:00Z               PEOPLE'S REPUBLIC OF CHINA   
8  2013-10-29T00:00:00Z                  THE GOVERNMENT OF INDIA   
9  2013-10-29T00:00:00Z                   THE KINGDOM OF MOROCCO   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   
5                   NaN                        Republic of Kenya!$!KE   
6  2019-06-30T00:00:00Z                        Republic of India!$!IN   
7                   NaN               People's Republic of China!$!CN   
8  2018-12-31T00:00:00Z                        Republic of India!$!IN   
9  2014-12-31T00:00:00Z                       Kingdom of Morocco!$!MA   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   
5          KE                        Republic of Kenya               Kenya   
6          IN                        Republic of India               India   
7          CN               People's Republic of China               China   
8          IN                        Republic of India               India   
9          MA                       Kingdom of Morocco             Morocco   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   
5                        ...                                   JB   IBRD   
6                        ...                                   TI   IBRD   
7                        ...                                   LR   IBRD   
8                        ...                                   TI   IBRD   
9      

In [12]:
# check column names
df.columns

Index([u'_id', u'approvalfy', u'board_approval_month', u'boardapprovaldate',
       u'borrower', u'closingdate', u'country_namecode', u'countrycode',
       u'countryname', u'countryshortname', u'docty',
       u'envassesmentcategorycode', u'grantamt', u'ibrdcommamt', u'id',
       u'idacommamt', u'impagency', u'lendinginstr', u'lendinginstrtype',
       u'lendprojectcost', u'majorsector_percent', u'mjsector_namecode',
       u'mjtheme', u'mjtheme_namecode', u'mjthemecode', u'prodline',
       u'prodlinetext', u'productlinetype', u'project_abstract',
       u'project_name', u'projectdocs', u'projectfinancialtype',
       u'projectstatusdisplay', u'regionname', u'sector', u'sector1',
       u'sector2', u'sector3', u'sector4', u'sector_namecode', u'sectorcode',
       u'source', u'status', u'supplementprojectflg', u'theme1',
       u'theme_namecode', u'themecode', u'totalamt', u'totalcommamt', u'url'],
      dtype='object')

The type of column name is unicode.

### 1. 10 countries with most projects

The relevant column names are country_namecode,	countrycode, countryname, and countryshortname. I will use countryshortname since it contains more familiar country names we use. 

In [13]:
country_names = df['countryshortname']

In [14]:
sum(country_names.isnull())

0

There is no missing country name.

In [15]:
# number of unique values
country_names.nunique()

118

There are 118 countries in toral.

In [16]:
# Country name and their frequency in projects sorted by frequency
project_count_by_country = country_names.value_counts()
project_count_by_country

China                               19
Indonesia                           19
Vietnam                             17
India                               16
Yemen, Republic of                  13
Morocco                             12
Bangladesh                          12
Nepal                               12
Africa                              11
Mozambique                          11
Brazil                               9
Burkina Faso                         9
Pakistan                             9
Armenia                              8
Tajikistan                           8
Tanzania                             8
Kyrgyz Republic                      7
Nigeria                              7
Lao People's Democratic Republic     7
Jordan                               7
Kenya                                6
West Bank and Gaza                   6
Afghanistan                          6
Peru                                 6
Nicaragua                            6
Congo, Democratic Republi

In [17]:
# Top 10 coutries
project_count_by_country[:10]

China                 19
Indonesia             19
Vietnam               17
India                 16
Yemen, Republic of    13
Morocco               12
Bangladesh            12
Nepal                 12
Africa                11
Mozambique            11
Name: countryshortname, dtype: int64

__Answer to question 1:__

The above output shows the 10 countries with most projects are __
China                 (19),
Indonesia             (19),
Vietnam               (17),
India                 (16),
Yemen, Republic of    (13),
Morocco               (12),
Bangladesh            (12),
Nepal                 (12),
Africa                (11),
Mozambique            (11) __
in order.
The numbers in parentheses are the number of proejects. 

### 2. Top 10 major project themes

The column __mjtheme_namecode__ will be used to find top major project themes.

In [18]:
# check the structure of mjtheme_namecode column
df['mjtheme_namecode'].head()

0    [{u'code': u'8', u'name': u'Human development'...
1    [{u'code': u'1', u'name': u'Economic managemen...
2    [{u'code': u'5', u'name': u'Trade and integrat...
3    [{u'code': u'7', u'name': u'Social dev/gender/...
4    [{u'code': u'5', u'name': u'Trade and integrat...
Name: mjtheme_namecode, dtype: object

In [19]:
# check more details of mjtheme_namecode for each of 10 projects
for item in df['mjtheme_namecode'][:10]:
    print(item) # themes for each project
    print(len(item)) # number of themes for each project

[{u'code': u'8', u'name': u'Human development'}, {u'code': u'11', u'name': u''}]
2
[{u'code': u'1', u'name': u'Economic management'}, {u'code': u'6', u'name': u'Social protection and risk management'}]
2
[{u'code': u'5', u'name': u'Trade and integration'}, {u'code': u'2', u'name': u'Public sector governance'}, {u'code': u'11', u'name': u'Environment and natural resources management'}, {u'code': u'6', u'name': u'Social protection and risk management'}]
4
[{u'code': u'7', u'name': u'Social dev/gender/inclusion'}, {u'code': u'7', u'name': u'Social dev/gender/inclusion'}]
2
[{u'code': u'5', u'name': u'Trade and integration'}, {u'code': u'4', u'name': u'Financial and private sector development'}]
2
[{u'code': u'6', u'name': u'Social protection and risk management'}, {u'code': u'6', u'name': u''}]
2
[{u'code': u'2', u'name': u'Public sector governance'}, {u'code': u'4', u'name': u'Financial and private sector development'}]
2
[{u'code': u'11', u'name': u'Environment and natural resources man

In [20]:
sum(df['mjtheme_namecode'].isnull())

0

In [21]:
df['mjtheme_namecode'][5][1]['name']

u''

There seems to be no missing value for the column mjtheme_namecode, but there are some blank theme names and redundant themes inside nested elements. The redundant themes could be there for some reasons (probably not by accident), so I will not remove them for now. The data also shows some projects have multiple themes, so each project has a list of dictionaries each with the theme code and name. Therefore, I will load json as a nested string this time and make tables using the techniques demonstrated above (normalization).

In [22]:
data = json.load(open('data/world_bank_projects.json'))

In [23]:
# Check the second project to see the structure
# data[1] 

I will use normalization to get a row with two columns, theme code and theme name. Each row is for each theme of every project. 

In [43]:
# normalization
df_theme = json_normalize(data, 'mjtheme_namecode')
df_theme.head(20)

code                                          name
0     8                             Human development
1    11                                              
2     1                           Economic management
3     6         Social protection and risk management
4     5                         Trade and integration
5     2                      Public sector governance
6    11  Environment and natural resources management
7     6         Social protection and risk management
8     7                   Social dev/gender/inclusion
9     7                   Social dev/gender/inclusion
10    5                         Trade and integration
11    4      Financial and private sector development
12    6         Social protection and risk management
13    6                                              
14    2                      Public sector governance
15    4      Financial and private sector development
16   11  Environment and natural resources management
17    8                                              
18   10                             Rural development
19    7

In [26]:
#sum(df_theme.name.isnull())
df_theme.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1499 entries, 0 to 1498
Data columns (total 2 columns):
code    1499 non-null object
name    1499 non-null object
dtypes: object(2)
memory usage: 35.1+ KB


In [27]:
len(df_theme[df_theme['code'] ==""])

0

In [28]:
len(df_theme[df_theme['name'] ==""])

122

There are no null objects for both code and name columns, but 122 names are missing. I will make a list of indices with missing names.

In [29]:
# Make a list of indices with missing names
no_name_idx = df_theme[df_theme['name']==""].index.tolist()
len(no_name_idx)

122

In [30]:
type(df_theme['code'][0]), type(df_theme['name'][0]), type(df_theme.columns[0])

(unicode, unicode, unicode)

Both columns and column names are unicode types, so I will make them integer, string, and string types, respectively.

In [31]:
df_theme['code'] = df_theme['code'].astype(int)
df_theme['name'] = df_theme['name'].astype(str)
df_theme.columns = df_theme.columns.astype(str)

type(df_theme['code'][0]), type(df_theme['name'][0]), type(df_theme.columns[0])

(numpy.int32, str, str)

In [32]:
df_theme['code'].value_counts()

11    250
10    216
8     210
2     199
6     168
4     146
7     130
5      77
9      50
1      38
3      15
Name: code, dtype: int64

In [33]:
code_count_table = pd.DataFrame(df_theme['code'].value_counts().reset_index())
code_count_table.columns =['code', 'count']
code_count_table

code  count
0     11    250
1     10    216
2      8    210
3      2    199
4      6    168
5      4    146
6      7    130
7      5     77
8      9     50
9      1     38
10     3     15

There are 11 major themes with code 1 to 11. I will find out the name for each theme code and make a table.

In [34]:
code_name_table = df_theme.drop(no_name_idx).drop_duplicates().sort_values(by='code').reset_index(drop=True)
code_name_table

code                                          name
0      1                           Economic management
1      2                      Public sector governance
2      3                                   Rule of law
3      4      Financial and private sector development
4      5                         Trade and integration
5      6         Social protection and risk management
6      7                   Social dev/gender/inclusion
7      8                             Human development
8      9                             Urban development
9     10                             Rural development
10    11  Environment and natural resources management

In [35]:
merge_on_code = pd.merge(code_count_table, code_name_table, on='code')
merge_on_code

code  count                                          name
0     11    250  Environment and natural resources management
1     10    216                             Rural development
2      8    210                             Human development
3      2    199                      Public sector governance
4      6    168         Social protection and risk management
5      4    146      Financial and private sector development
6      7    130                   Social dev/gender/inclusion
7      5     77                         Trade and integration
8      9     50                             Urban development
9      1     38                           Economic management
10     3     15                                   Rule of law

In [36]:
# Top 10 major themes
merge_on_code[['code','name']][:10]

code                                          name
0    11  Environment and natural resources management
1    10                             Rural development
2     8                             Human development
3     2                      Public sector governance
4     6         Social protection and risk management
5     4      Financial and private sector development
6     7                   Social dev/gender/inclusion
7     5                         Trade and integration
8     9                             Urban development
9     1                           Economic management

__Answer to question 2:__

The above table shows codes and names of the top 10 major themes.

### 3. Fill missing names

In [37]:
df = pd.read_json('data/world_bank_projects.json')

In [38]:
df['mjtheme_namecode'].head()

0    [{u'code': u'8', u'name': u'Human development'...
1    [{u'code': u'1', u'name': u'Economic managemen...
2    [{u'code': u'5', u'name': u'Trade and integrat...
3    [{u'code': u'7', u'name': u'Social dev/gender/...
4    [{u'code': u'5', u'name': u'Trade and integrat...
Name: mjtheme_namecode, dtype: object

In [39]:
df['mjtheme_namecode'][0]

[{u'code': u'8', u'name': u'Human development'},
 {u'code': u'11', u'name': u''}]

In [40]:
mjtheme_dict_unicode = dict(zip(code_name_table['code'].astype(unicode),code_name_table['name'].astype(unicode)))
mjtheme_dict_unicode

{u'1': u'Economic management',
 u'10': u'Rural development',
 u'11': u'Environment and natural resources management',
 u'2': u'Public sector governance',
 u'3': u'Rule of law',
 u'4': u'Financial and private sector development',
 u'5': u'Trade and integration',
 u'6': u'Social protection and risk management',
 u'7': u'Social dev/gender/inclusion',
 u'8': u'Human development',
 u'9': u'Urban development'}

In [41]:
count_fixed=0
for i, project in enumerate(df['mjtheme_namecode']):
    for j, theme in enumerate(project):
        if theme['name'] == "":
            print "Before: ", df['mjtheme_namecode'][i][j]
            df['mjtheme_namecode'][i][j]['name'] = mjtheme_dict_unicode[theme['code']]
            print "After: ", df['mjtheme_namecode'][i][j]
            count_fixed += 1

print "In total, ", count_fixed, " names were filled!!"

Before:  {u'code': u'11', u'name': u''}
After:  {u'code': u'11', u'name': u'Environment and natural resources management'}
Before:  {u'code': u'6', u'name': u''}
After:  {u'code': u'6', u'name': u'Social protection and risk management'}
Before:  {u'code': u'8', u'name': u''}
After:  {u'code': u'8', u'name': u'Human development'}
Before:  {u'code': u'7', u'name': u''}
After:  {u'code': u'7', u'name': u'Social dev/gender/inclusion'}
Before:  {u'code': u'2', u'name': u''}
After:  {u'code': u'2', u'name': u'Public sector governance'}
Before:  {u'code': u'4', u'name': u''}
After:  {u'code': u'4', u'name': u'Financial and private sector development'}
Before:  {u'code': u'2', u'name': u''}
After:  {u'code': u'2', u'name': u'Public sector governance'}
Before:  {u'code': u'7', u'name': u''}
After:  {u'code': u'7', u'name': u'Social dev/gender/inclusion'}
Before:  {u'code': u'4', u'name': u''}
After:  {u'code': u'4', u'name': u'Financial and private sector development'}
Before:  {u'code': u'11',

In [42]:
df['mjtheme_namecode'][0]

[{u'code': u'8', u'name': u'Human development'},
 {u'code': u'11', u'name': u'Environment and natural resources management'}]

__Answer to question 3:__

In total, 122 theme names were filled.